hop/io.py

```diff
@@ -396,29 +396,75 @@ def write(self, message, headers=None, delivery_callback=errors.raise_delivery_e
                 is either delivered or permenantly fails to be delivered.
 
         """
-        self._producer.write(self._pack(message), headers=headers,
-                             delivery_callback=delivery_callback)
+        message, headers = self.pack(message, headers)
+        self._producer.write(message, headers=headers, delivery_callback=delivery_callback)
+
+    def write_raw(self, packed_message, headers=None,
+                  delivery_callback=errors.raise_delivery_errors):
+        """Write a pre-encoded message to the stream.
+
+        This is an advanced interface; for most purposes it is preferrable to use
+        :meth:`Producer.write <hop.io.Producer.write>` instead.
+
+        Args:
+            packed_message: The message to write, which must already be correctly encoded by
+                            :meth:`Producer.pack <hop.io.Producer.pack>`
+            headers: Any headers to attach to the message, either as a dictionary
+                mapping strings to strings, or as a list of 2-tuples of strings.
+            delivery_callback: A callback which will be called when each message
+                is either delivered or permenantly fails to be delivered.
+
+        """
+        self._producer.write(packed_message, headers=headers, delivery_callback=delivery_callback)
 
     @staticmethod
-    def _pack(message):
-        """Pack and serialize messages.
+    def pack(message, headers=None):
+        """Pack and serialize a message.
+
+        This is an advanced interface, which most users should not need to call directly, as
+        :meth:`Producer.write <hop.io.Producer.write>` uses it automatically.
 
         Args:
             message: The message to pack and serialize.
+            headers: The set of headers requested to be sent with the message, either as a
+                     dictionary mapping strings to strings, or as a list of 2-tuples of strings.
+
+        Returns: A tuple containing the serialized message and the collection of headers which
+                 should be sent with it.
 
-        :meta private:
         """
+        # canonicalize headers to list form
+        if headers is None:
+            headers = []
+        if type(headers) == dict:
+            headers = list(headers.items())
+
+        # ensure all headers are encoded
+        def ensure_bytes_like(thing):
+            try:  # check whether thing is bytes-like
+                memoryview(thing)
+                return thing  # keep as-is
+            except TypeError:
+                return thing.encode("utf-8")
```

Comment:
Personally, I think adding the function inline is less readable than if we moved this to a private function in the same module.

Regarding the content of the function, how about adding an `else` to the try/except?
```
            try:  # check whether thing is bytes-like
                memoryview(thing)
            except TypeError:
                return thing.encode("utf-8")
            else:
                return thing 
```

Reply:
I personally find it more readable not to have things like this floating around at the module scope when they're only intended for use in one place, but I also don't care strongly.